career-mentorship-program/2021-spring/matching_code.py

```diff
@@ -410,9 +503,66 @@ def compute_weight_and_statistics(self):
         if year_difference == 1 or year_difference == 2:
             weight += MENTOR_CLOSE_IN_YEAR_BONUS
 
-
         """ !!! Add any new matching criteria here !!! """
 
+        # 4. Hobbies
+
+        num_hobbies_overlap = len(self.mentee.hobbies & self.mentor.hobbies)
+        num_mentee_hobbies = len(self.mentee.hobbies)
+        hobbies_overlap_fraction = num_hobbies_overlap / num_mentee_hobbies
+        statistics["hobbies_overlap_fraction"] = hobbies_overlap_fraction
+        weight += HOBBIES_TOPICS_MULTIPLIER * hobbies_overlap_fraction
+
+        # function for gender and race
+        def gender_and_race(mentor_preference, mentee_preference, stat_field1, stat_field2, bonus):
+            if mentor_preference and mentee_preference:
+                does_prefer = True
+                statistics[stat_field1] = True
+            else:
+                does_prefer = False
+                statistics[stat_field1] = False
+            if does_prefer:
+                preferred_and_same = self.mentor.actual == self.mentee.actual
+                statistics[stat_field2] = preferred_and_same
+            else:
+                preferred_and_same = False
+                statistics[stat_field2] = preferred_and_same
+            if does_prefer and preferred_and_same:
+                self.weight += bonus
+
+        # 5. Gender
+        # checks if both want to be matched with same gender and are same gender, if so add bonus
+        gender_and_race(self.mentor.does_prefer_gender, self.mentee.does_prefer_gender, 'does_prefer_gender',
+                        'gender_preferred_and_same', GENDER_SAME_BONUS)
+
+        # 6. Race
+        # checks if both want to be matched with same race and are same race, if so add bonus
+        gender_and_race(self.mentor.does_prefer_race, self.mentee.does_prefer_race, 'does_prefer_race',
+                        'race_preferred_and_same', RACE_SAME_BONUS)
+
+        # 7. Time Zone
+        mentor_time_zone = self.mentor.time_zone
+        mentee_time_zone = self.mentee.time_zone
+        time_difference = abs(mentor_time_zone - mentee_time_zone)
+        statistics['time_difference'] = time_difference
+        if time_difference >= 5:
+            weight -= TIME_ZONE_DIFFERENCE_PENALTY
+
+        # 8. Time available
+        mentor_time_available = self.mentor.time_available
+        mentee_time_available = self.mentee.time_available
+        time_available = abs(mentor_time_available - mentee_time_available)
+        statistics['time_available'] = time_available
+        if time_available > 2:
+            weight -= TIME_AVAILABLE_DIFFERENCE_PENALTY
+
+        # 9. CS experience
+        mentor_cs_experience = self.mentor.cs_experience
+        mentee_cs_experience = self.mentee.cs_experience
+        cs_experience_difference = mentor_cs_experience - mentee_cs_experience
+        statistics['cs_experience_difference'] = cs_experience_difference
+        if cs_experience_difference < 0:
```

Comment:
Note that the form has the mentor cs experience on a scale of 1-3 and the mentee experience on a scale of 1-5. This means there will be many cases where the `cs_experience_difference` is actually less than zero just because of how we designed the scales. We could either rework the logic here or just change the form to use a scale of 1-3 for both mentees and mentors.